### PR TITLE
973 pack to one preference

### LINKED ISSUE #973: Should we handle the "pack to one" preference?
mSupply has a store preference "Set pack to one for all visible items in the store". When a store with this preference turned on receives stock, the system automatically sets this stock to have pack sizes of one (and does the necessary conversions for packs, quantity, price, etc). 

This is primarily used for facilities that are either smaller, or dispense to patients, and therefore mainly work in pack sizes of 1. It
is widely used. 

The alternative to this functionality is being able to repack items manually to packs of 1 when received. The caution with this approach is that it may require some training and monitoring - staff who are used to always handling packs of one may accidentally issue out large amounts of larger packsizes. 

image

My questions are: 
- Around how long would this work take?
- Do we have a design for dealing with store prefs?
- Is it planned to deal with store prefs in omSupply at all? 
- Is it better to do the manual repack function?

### PR DESCRIPTION
**Pull request by roxy-dao**
Closes #973

- Inbound shipment lines will recalculate relevant fields to be in packs of one 
- Relevant fields will also be recalculated when an outbound shipment is sent to a store with pack to one preference turned on.

### Inbound shipment manual testing
- [ ] Turn on pack to one preference on central server for store
- [ ] Go to store with pack to one preference and create an inbound shipment
- [ ] Add line with pack size > 1 
- [ ] Click OK - see changes in line

### Outbound shipment manual testing
- [ ] Follow steps 1-3 above
- [ ] Change status to picked or shipped
- [ ] Go to destination store 
- [ ] See recalculated invoice

### CODE REVIEW

File: client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/TabTables.tsx [hunk 1]
```diff
@@ -137,7 +137,7 @@ export const PricingTableComponent: FC<TableProps> = ({
         'lineTotal',
         {
           accessor: ({ rowData }) =>
-            rowData.numberOfPacks * rowData.packSize * rowData.costPricePerPack,
+            rowData.numberOfPacks * rowData.costPricePerPack,
```
Comment (roxy-dao) [inline client/packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/TabTables.tsx:140]: 🤔 Miscalculated line total

**Review by andreievg — CHANGES_REQUESTED** [review]
Although this looks very good, I think it's better to have a `transformer` type function for stock line and invoice line:

* don't have to use optional input for generate_batch
* organises logic by concern

i.e. 

```rust
fn convert_stock_line_to_single_pack(line: StockLineRow) -> StockLineRow {
   StockLineRow {
        number_of_packs = number_of_packs * line.pack_size as f64,
        sell_price_per_pack = sell_price_per_pack / line.pack_size as f64,
        cost_price_per_pack = cost_price_per_pack / line.pack_size as f64,
        pack_size: 1
     ...line
    }
}

// ....
        let new_batch = generate_batch(
            &existing_invoice_row.store_id,
            new_line.clone(),
            false,
            &existing_invoice_row.name_id,
        );

        let new_batch = match store_preferences.pack_to_one {
            true => convert_stock_line_to_single_pack(new_batch),
            false => new_batch,
        };
```

**Comment by roxy-dao** [comment]
> d45bced

File: server/service/src/invoice_line/inbound_shipment_line/generate.rs [hunk 2]
```diff
@@ -1,6 +1,20 @@
 use repository::{InvoiceLineRow, StockLineRow};
 use util::uuid::uuid;
 
+pub fn convert_stock_line_to_single_pack(
+    stock_line: StockLineRow,
+    invoice_line: &InvoiceLineRow,
```
Comment (andreievg) [inline server/service/src/invoice_line/inbound_shipment_line/generate.rs:6]: Hmm any reason to pass through invoice_line, rather then just do it from stock_line ?

File: server/service/src/processors/transfer/shipment/common.rs [hunk 3]
```diff
@@ -25,18 +28,25 @@ pub(crate) fn generate_inbound_shipment_lines(
...
-                 number_of_packs,
+                 mut number_of_packs,
                  note,
                  r#type,
                  total_after_tax: _,
                  total_before_tax: _,
                  tax,
                  inventory_adjustment_reason_id: _,
              }| {
-                let cost_price_per_pack = sell_price_per_pack;
```
Comment (andreievg) [inline server/service/src/processors/transfer/shipment/common.rs:39]: I think it's a good idea to be `dry` when logic is split between context (i.e. define pack to one conversion in one place)

```rust
  let lines_iter = outbound_lines.into_iter().map(|l| l.invoice_line_row);

    let lines_iter = match store_preferences.pack_to_one {
        true => lines_iter.map(convert_invoice_line_to_single_pack),
        false => lines_iter,
    };

    let inbound_lines = lines_iter
        .map(
            |InvoiceLineRow {
```

**Review by andreievg — COMMENTED** [review]
do we need to consider inbound shipment line update ?

File: server/service/src/invoice_line/inbound_shipment_line/generate.rs (hunk 2 above)
Comment (roxy-dao) [reply server/service/src/invoice_line/inbound_shipment_line/generate.rs:6]: Nope don't need you're correct (:

**Comment by roxy-dao** [comment]
> do we need to consider inbound shipment line update ?

The insert should convert everything to one already, so don't we need to consider inbound shipment line update... but can add it in to be safe if you want~

File: server/service/src/processors/transfer/shipment/common.rs (hunk 3 above)
Comment (roxy-dao) [reply server/service/src/processors/transfer/shipment/common.rs:39]: [6e56076](https://github.com/openmsupply/open-msupply/pull/1167/commits/6e56076c3c08b3a142d5510183baca26f7437df4)

**Comment by andreievg** [comment]
> The insert should convert everything to one already, so don't we need to consider inbound shipment line update... but can add it in to be safe if you want~

I think since pack size can be edited in update we should also do this in updated (should be quite easy with the helper method you added)

So yes please can you add it to update logic too

**Comment by roxy-dao** [comment]
> So yes please can you add it to update logic too

Ah right! Still being able to edit pack size slipped my mind 🙇🏻 
[c93cc35](https://github.com/openmsupply/open-msupply/pull/1167/commits/c93cc351ade38a6f5d0cf73bcdc5f42ee46cdf71)

**Review by andreievg — APPROVED** [review]
🙏 looks great

**Pull request merged**